modPrecompiled/src/org/aion/precompiled/contracts/EDVerifyContract.java

```diff
@@ -0,0 +1,74 @@
+package org.aion.precompiled.contracts;
+
+import org.aion.base.type.Address;
+import org.aion.base.type.IExecutionResult;
+import org.aion.crypto.ed25519.ECKeyEd25519;
+import org.aion.crypto.ed25519.Ed25519Signature;
+import org.aion.vm.ExecutionResult;
+import org.aion.vm.IPrecompiledContract;
+
+import java.util.Arrays;
+
+/**
+ * Precompiled contract that verifies a signed message with the Ed25519 signature algorithm
+ */
+public class EDVerifyContract implements IPrecompiledContract {
+    // set to a default cost for now, this will need to be adjusted
+    private final static long COST = 21000L;
+
+    /**
+     * <p>
+     * Method takes as input a byte array representing the edverify method parameters represented in hexadecimal
+     * from the solidity binary as follows
+     * <p>
+     * edverify(bytes32 hash, bytes32 public_key, bytes signature)
+     * <p>
+     * input
+     * |
+     * | message hash (32 bytes) | public key (32 bytes) | signature bytes (64 bytes) |
+     * <p>
+     * where
+     * - message hash = keccak 256 hashed message
+     * - public key = public key of the ED25519 key pair
+     * - signature = 64 bytes of the ED25519 signature
+     *
+     * @param input    The input arguments for the contract.
+     * @param nrgLimit The energy limit.
+     * @return byte array of length 32 containing: the address of the account that signed the message  if the message was
+     *                                              signed with the public key of the account sent as a parameter
+     * <p>
+     *                                              ZERO_ADDRESS 32 byte array of 0 otherwise
+     */
+    @Override
+    public IExecutionResult execute(byte[] input, long nrgLimit) {
+        if (COST > nrgLimit) {
+            return new ExecutionResult(ExecutionResult.ResultCode.OUT_OF_NRG, 0, Address.ZERO_ADDRESS().toBytes());
+        }
+        byte[] msg = new byte[32];
+        byte[] pubKey = new byte[32];
+        byte[] sig = new byte[64];
+
+        try {
```

Comment:
All LGTM. Just wondering, is it worth it to return an error if the input length is larger than expected, since this may indicate that the user has incorrectly understood the format?

Reply:
@aionick What do you think about creating a new `ResultCode` like `ExecutionResult.ResultCode.INVALID_INPUT`? Are these codes documented for the clients that use the precompiled contracts? So I know if I need to update the docs.

Reply:
I'd like to look into this. From my understanding `ResultCode` is only used internally and never makes it out to the user. I may be wrong here. But I do think we need a little more flexibility in the precompileds for documenting errors. @yao-sun @AionJayT what do you guys think?

Reply:
@centrys-david, we can create the case to handle the INVALID_INPUT. But it requires a few places change.
@aionick we definitely need to define how to use return code in case the developer misusing.

Reply:
@AionJayT @aionick 

There are 2 cases here:

1. Smart contract client
I don't know how to throw/return an error from the precompiled contract that can be interpreted by the aion_fastvm and solidity compiler. The best practices are using `require` so there needs to be a documentation for the client to say something like:
```
...
require(hash.length == 32)
require(pub.lenght == 32)
...
edveirfy(hash, pub, sig1, sig2)
...
```

2. Java client 
If someone wants to integrate this precompiled contract on a lower level (eg. calling it directly) there is a minium amount of logging [here](https://github.com/aionnetwork/aion/pull/589/files/bcac62e015b7f595e713fbe04ed1d6a04d0198b7#diff-ae8e32d7dc609cad8c9618627777a278R57) that one can check on the kernel when developing. 

Still, I think documentation is our best choice for now and a new `ExecutionResult.ResultCode.INVALID_INPUT`

Let me know what you think.